plugins/dell-dock2/fu-dell-dock2-ec.c

```diff
@@ -0,0 +1,972 @@
+}
+
+static gboolean
+fu_dell_dock2_ec_reload(FuDevice *device, GError **error)
+{
+	/* if query looks bad, wait a few seconds and retry */
+	if (!fu_device_retry_full(device, fu_dell_dock2_ec_query_cb, 10, 2000, NULL, error)) {
+		g_prefix_error(error, "failed to query dock ec: ");
+		return FALSE;
+	}
+	return TRUE;
+}
+
+static gboolean
+fu_dell_dock2_ec_setup(FuDevice *device, GError **error)
+{
+	guint32 ec_version = 0;
+
+	/* FuUsbDevice->setup */
+	if (!FU_DEVICE_CLASS(fu_dell_dock2_ec_parent_class)->setup(device, error))
+		return FALSE;
+
+	/* get dock type */
+	if (!fu_dell_dock2_ec_dock_type_cmd(device, error))
+		return FALSE;
+
+	/* if query looks bad, wait a few seconds and retry */
+	if (!fu_device_retry_full(device, fu_dell_dock2_ec_query_cb, 10, 2000, NULL, error)) {
+		g_prefix_error(error, "failed to query dock ec: ");
+		return FALSE;
+	}
+
+	/* setup version */
+	ec_version = GUINT32_FROM_BE(fu_dell_dock2_ec_get_ec_version(device));
+	fu_device_set_version_raw(device, ec_version);
+
+	/* create the subcomponents */
+	if (!fu_dell_dock2_ec_probe_subcomponents(device, error))
+		return FALSE;
+
+	g_debug("dell-dock2-ec->setup done successfully");
+	return TRUE;
+}
+
+static gchar *
+fu_dell_dock2_ec_convert_version(FuDevice *device, guint64 version_raw)
+{
+	return fu_hex_version_from_uint32(version_raw, fu_device_get_version_format(device));
+}
+
+static gboolean
+fu_dell_dock2_ec_open(FuDevice *device, GError **error)
+{
+	/* FuUdevDevice->open */
+	return FU_DEVICE_CLASS(fu_dell_dock2_ec_parent_class)->open(device, error);
+}
+
+static void
+fu_dell_dock2_ec_finalize(GObject *object)
+{
+	FuDellDock2Ec *self = FU_DELL_DOCK2_EC(object);
+	g_free(self->dock_data);
+	g_free(self->dock_info);
+	g_free(self->raw_versions);
+	g_free(self->ec_minimum_version);
+	G_OBJECT_CLASS(fu_dell_dock2_ec_parent_class)->finalize(object);
+}
+
+static void
+fu_dell_dock2_ec_set_progress(FuDevice *self, FuProgress *progress)
+{
+	fu_progress_set_id(progress, G_STRLOC);
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_RESTART, 0, "detach");
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_WRITE, 100, "write");
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_RESTART, 0, "attach");
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_BUSY, 0, "reload");
+}
+
+static void
+fu_dell_dock2_ec_init(FuDellDock2Ec *self)
+{
+	self->dock_data = g_new0(FuDellDock2DockDataStructure, 1);
+	self->raw_versions = g_new0(FuDellDock2DockFWVersion, 1);
+	self->dock_info = g_new0(FuDellDock2DockInfoStructure, 1);
+
+	fu_device_add_protocol(FU_DEVICE(self), "com.dell.dock2");
+	fu_device_add_vendor_id(FU_DEVICE(self), "USB:0x413C");
+	fu_device_add_icon(FU_DEVICE(self), "dock-usb");
+	fu_device_set_summary(FU_DEVICE(self), "Dell Dock");
+	fu_device_add_flag(FU_DEVICE(self), FWUPD_DEVICE_FLAG_UPDATABLE);
+	fu_device_add_flag(FU_DEVICE(self), FWUPD_DEVICE_FLAG_SIGNED_PAYLOAD);
+	fu_device_add_flag(FU_DEVICE(self), FWUPD_DEVICE_FLAG_USABLE_DURING_UPDATE);
```

Comment:
shouldn't this only be set when passive flow is in use?

Reply:
default with passive flow, will change to optional when online activation become possible.